// @vtex/api/lib/ID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HttpClient_1 = require("./HttpClient");
const routes = {
    SEND: '/accesskey/send',
    START: '/start',
    VALIDATE: '/accesskey/validate',
    VALIDATE_CLASSIC: '/classic/validate',
};
const VTEXID_ENDPOINTS = {
    BETA: 'https://vtexid.vtex.com.br/api/vtexid/pub/authentication',
    STABLE: 'https://vtexid.vtex.com.br/api/vtexid/pub/authentication',
};
const endpoint = (env) => {
    return VTEXID_ENDPOINTS[env] || env;
};
class ID {
    constructor(endpointUrl = 'STABLE', opts) {
        this.getTemporaryToken = () => {
            return this.http.get(routes.START).then(({ authenticationToken }) => authenticationToken);
        };
        this.sendCodeToEmail = (token, email) => {
            const params = { authenticationToken: token, email };
            return this.http.get(routes.SEND, { params, headers: this.defaultHeaders });
        };
        this.getEmailCodeAuthenticationToken = (token, email, code) => {
            const params = {
                accesskey: code,
                authenticationToken: token,
                login: email,
            };
            return this.http.get(routes.VALIDATE, { params, headers: this.defaultHeaders });
        };
        this.getPasswordAuthenticationToken = (token, email, password) => {
            const params = {
                authenticationToken: token,
                login: email,
                password,
            };
            return this.http.get(routes.VALIDATE_CLASSIC, { params, headers: this.defaultHeaders });
        };
        this.http = HttpClient_1.HttpClient.forLegacy(endpoint(endpointUrl), opts);
        this.defaultHeaders = opts.accept ? { accept: opts.accept } : {};
    }
}
exports.ID = ID;
