// @vtex/api/lib/Workspaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
const IODataSource_1 = require("./IODataSource");
const routes = {
    Account: (account) => `/${account}`,
    Promote: (account) => `${routes.Workspace(account, constants_1.DEFAULT_WORKSPACE)}/_promote`,
    Workspace: (account, workspace) => `${routes.Account(account)}/${workspace}`,
};
class Workspaces extends IODataSource_1.IODataSource {
    constructor() {
        super(...arguments);
        this.service = 'router';
        this.httpClientFactory = IODataSource_1.forRoot;
        this.list = (account) => {
            return this.http.get(routes.Account(account));
        };
        this.get = (account, workspace) => {
            return this.http.get(routes.Workspace(account, workspace));
        };
        this.set = (account, workspace, metadata) => {
            return this.http.put(routes.Workspace(account, workspace), metadata);
        };
        this.create = (account, workspace) => {
            return this.http.post(routes.Account(account), { name: workspace });
        };
        this.delete = (account, workspace) => {
            return this.http.delete(routes.Workspace(account, workspace));
        };
        this.reset = (account, workspace, metadata = {}) => {
            const params = { reset: true };
            return this.http.put(routes.Workspace(account, workspace), metadata, { params });
        };
        this.promote = (account, workspace) => {
            return this.http.put(routes.Promote(account), { workspace });
        };
    }
}
exports.Workspaces = Workspaces;
