// @vtex/api/lib/caches/DiskCache.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const ReadWriteLock = require("rwlock");
class DiskCache {
    constructor(cachePath, readFile = fs_extra_1.readJSON, writeFile = fs_extra_1.outputJSON) {
        this.cachePath = cachePath;
        this.readFile = readFile;
        this.writeFile = writeFile;
        this.hits = 0;
        this.total = 0;
        this.has = (key) => {
            const pathKey = this.getPathKey(key);
            return fs_extra_1.pathExistsSync(pathKey);
        };
        this.getStats = (name = 'disk-cache') => {
            const stats = {
                hits: this.hits,
                name,
                total: this.total,
            };
            this.hits = 0;
            this.total = 0;
            return stats;
        };
        this.get = (key) => __awaiter(this, void 0, void 0, function* () {
            const pathKey = this.getPathKey(key);
            this.total += 1;
            const data = yield new Promise(resolve => {
                this.lock.readLock(key, (release) => __awaiter(this, void 0, void 0, function* () {
                    try {
                        const fileData = yield this.readFile(pathKey);
                        release();
                        this.hits += 1;
                        resolve(fileData);
                    }
                    catch (e) {
                        release();
                        resolve(undefined);
                    }
                }));
            });
            return data;
        });
        this.set = (key, value) => __awaiter(this, void 0, void 0, function* () {
            const pathKey = this.getPathKey(key);
            const failure = yield new Promise(resolve => {
                this.lock.writeLock(key, (release) => __awaiter(this, void 0, void 0, function* () {
                    try {
                        const writePromise = yield this.writeFile(pathKey, value);
                        release();
                        resolve(writePromise);
                    }
                    catch (e) {
                        release();
                        resolve(true);
                    }
                }));
            });
            return !failure;
        });
        this.getPathKey = (key) => path_1.join(this.cachePath, key);
        this.lock = new ReadWriteLock();
    }
}
exports.DiskCache = DiskCache;
